pkgs/dart_model/lib/src/json_changes.dart

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) 2024, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'json.dart';
+
+/// The changes between two [Json] instances.
+extension type JsonChanges.fromJson(Map<String, Object?> node) {
+  JsonChanges({
+    required List<Update> updates,
+    required List<Removal> removals,
+  }) : this.fromJson({
+          'updates': updates,
+          'removals': removals,
+        });
+
+  /// The nodes that were updated by these changes.
+  List<Update> get updates => (node['updates']! as List).cast();
+
+  /// The nodes that were removed by these changes.
+  List<Removal> get removals => (node['removals']! as List).cast();
+}
+
+/// A path into JSON data.
+extension type Path.fromJson(List<Object?> node) {
```

Comment:
These paths are extremely expensive. They should be represented as linked lists or something? I am not sure how exactly to translate that into JSON in the best way, but I think it is critical that we don't duplicate the entire path for each update, copying the lists etc.

Reply:
Yeah this was painful to write :)

I don't want to optimize prematurely though, I think a TODO is the right thing for now, it looks bad but it's not clear it's a hotspot for performance, and it should be possible to optimize in place if so.

One thing that does need figuring out is what we are doing here with mutability, if we have a good story that can help a lot, let me TODO that one too :)

The translation to (binary) JSON should be fine, I am 99% sure we will end up deduping key strings so they are small and fast, we can treat strings in paths the same.

Reply:
Paths are gone following the change to maps :)

That's not the end of the performance questions, though. Right now both creating the diff and applying changes do a lot of instantiation/copying. I think it's possible to improve that, but it's a lot more complexity than fits in this PR :)